### DISCUSSION TITLE
Why is the codebase not using environemnt variables from `.env` files

### DISCUSSION THREAD

**Discussion by divyanshchahar** [Infisical/infisical#2210] [Q&A]
The code does not makes use of any of the environemnt variables from `.env` file. This is because one of the following reasons :

- Most of the envrinment variables are provided by the `getConfig` function located in `backend/src/lib/config/env.ts`. However, since `backend/src/main.ts` does not configures `dotenv` correctly ( `dotenv` needs to have a `path` attribute specified to locate `.env`, since it is not located at the same level), `backend/src/lib/config/env.ts` cannot locate the `.env`.
- Wherever envoronment variables are used without `getConfig` function ( for example in `backend/src/db/knexfile.ts` ), they are using `dotnev` to acess these environment variables, but `dotenv` in those files is either missing `path` attribute or the path resolution cannot happen properly.

Below is a detailed description for every environment variable and why they are essentially unused in the code base

# Provided by `getConfig` in `backend/src/lib/config/env.ts` but not read from `.env`

All the environemnt variables listed in this section are not read from the `.env` file because `process.env.ENVIRONMENT_VARIABLE` is not used in the `getConfig` function

- `AUTH_SECRET`
- `REDIS_URL`
- `SITE_URL`
- `SMTP_HOST`
- `SMTP_USERNAME`
- `SMTP_PASSWORD`
- `CLIENT_ID_HEROKU`
- `CLIENT_ID_VERCEL`
- `CLIENT_ID_NETLIFY`
- `CLIENT_ID_GITHUB`
- `CLIENT_ID_GITLAB`
- `CLIENT_ID_BITBUCKET`
- `CLIENT_SECRET_HEROKU`
- `CLIENT_SECRET_VERCEL`
- `CLIENT_SECRET_NETLIFY`
- `CLIENT_SECRET_GITHUB`
- `CLIENT_SECRET_GITLAB`
- `CLIENT_SECRET_BITBUCKET`
- `CLIENT_SLUG_VERCEL`
- `CLIENT_ID_GOOGLE_LOGIN`
- `CLIENT_SECRET_GOOGLE_LOGIN`
- `CLIENT_ID_GITHUB_LOGIN`
- `CLIENT_SECRET_GITHUB_LOGIN`
- `CLIENT_ID_GITLAB_LOGIN`
- `CLIENT_SECRET_GITLAB_LOGIN`
- `CAPTCHA_SECRET`
- `PLAIN_API_KEY`
- `PLAIN_WISH_LABEL_IDS`

# Provided by `getConfig` in `backend/src/lib/config/env.ts` but values are hardcoded

All the values in this section are also provided by `getConfig` but instead of being read from the `.env` file they are hardocded

- `SMTP_PORT`
- `POSTHOG_HOST`
- `POSTHOG_PROJECT_API_KEY`

# Declared in `.env` but not used anywhere in the codebase

- `POSTGRES_PASSWORD`
- `POSTGRES_USER`
- `POSTGRES_DB`
- `SMTP_NAME`
- `SENTRY_DSN`

# Other issues

- `DB_CONNECTION_URI` : This environment variable is provided by `getConfig` and is also directly used in `backend/src/db/knexfile.ts`. Since `dotenv` is not configured with `path` attribute in `backend/src/db/knexfile.ts` is unable to locate the `.env`. In the `getConfig` function the value is not read from `.env`

**Comment by maidul98** [comment]
The .env file is pased via Zod here https://github.com/Infisical/infisical/blob/main/backend/src/lib/config/env.ts#L167

And it is used throughout the app via https://github.com/Infisical/infisical/blob/main/backend/src/lib/config/env.ts#L164

**Reply by divyanshchahar** [reply]
- `process.env.ENV_VAR` is **rarely used** in any variable like it is done [here](https://github.com/Infisical/infisical/blob/46ca3856b3b0cf2410ac353e0c35fd9f2727ed7b/backend/src/lib/config/env.ts#L34) 

- `dotenv.config` is missing a `path` property [here](https://github.com/Infisical/infisical/blob/46ca3856b3b0cf2410ac353e0c35fd9f2727ed7b/backend/src/main.ts#L14) which is required whenever `.env` files are not in the same directory as the file which executes `dotenv.config`

- even the values which uses `process.env.ENV_VAR` like shown [here](https://github.com/Infisical/infisical/blob/46ca3856b3b0cf2410ac353e0c35fd9f2727ed7b/backend/src/lib/config/env.ts#L34) are undefined at runtime because `dotenv.config` in this [file](https://github.com/Infisical/infisical/blob/46ca3856b3b0cf2410ac353e0c35fd9f2727ed7b/backend/src/main.ts#L14) cannot resolve the path to `.env` file

- thus `.env` files are redundant at this point
